### ISSUE TITLE
EPOLLHUP or EPOLLRDHUP during connection establishment leads to a wrong decision that a backend is down

### ISSUE THREAD

**Issue by gbrdead** [haproxy/haproxy#1863]
### Detailed Description of the Problem

Sometimes (quite frequently under load) haproxy marks a backend as being down even though the liveness check (httpchk) is successful.

### Expected Behavior

Backends should not be marked as down if they aren't really down.

### Steps to Reproduce the Behavior

It is easy for me to reproduce it - just start about 10 simultaneous curl clients to pass requests continuously through haproxy at localhost. I have the feeling that this won't be easy for everybody, though...

### Do you have any idea what may have caused this?

1. After connect() is called on the socket that will be used for the liveness check, the socket file descriptor is added to an epoll descriptor. Occasionally (but frequently enough under load) epoll_wait() returns the following two events for the socket fd before the socket gets connected:
EPOLLIN|EPOLLHUP|EPOLLRDHUP
EPOLLIN|EPOLLRDHUP
These two events are always returned by the same call to epoll_wait() and always in this order (as witnessed by strace).

2. As a result of EPOLLHUP, _do_poll() in ev_epoll.c calls fd_update_events() with FD_EV_SHUT_W set. In the latter, the condition below the following comment:
/* SHUTW reported while FD was active for writes is an error */
is true, and FD_POLL_ERR gets set in fdtab[fd].state.
This flag never gets reset.

3. I have no idea what this EPOLLHUP means for a not-yet-connected socket, but there is a comment in sock_conn_check() in sock.c that suggest that this is to be expected. The comment is long and starts with:
/* Here we have 2 cases :
Since FD_POLL_ERR is set, connect() gets called again. It returns 0 and the connection to the backend is established.

4. The liveness check proceeds as usual - the request is sent and the response is received (as witnessed by strace).

5. Eventually, raw_sock_to_buf() in raw_sock.c gets called. Check its very end, after the label read0 and remember that FD_POLL_ERR is still set in fdtab[conn->handle.fd].state. conn->flags gets CO_FL_ERROR set and this leads to the liveness check to be considered as failed.

### Do you have an idea how to solve the issue?

I suspect that the trouble comes from FD_POLL_ERR not being reset before the second call to connect(). If I put the following line:
fdtab[fd].state &= ~(FD_POLL_ERR|FD_POLL_HUP);
in sock_conn_check() after the following comment:
/* error present, fall through common error check path */
the issue seems to get solved.

### What is your configuration?

```haproxy
global
    maxconn 50000

defaults
    log stdout local0 notice
    mode tcp
    maxconn 50000
    option redispatch
    option tcpka
    balance roundrobin
    default-server inter 2s fall 2 rise 2
    timeout connect 5s
    timeout client 50s
    timeout server 50s
    timeout tunnel 12h
    timeout http-keep-alive 1s
    timeout http-request 15s
    timeout queue 30s
    timeout tarpit 60s

backend bbb
    mode http
    option httpchk GET /liveness
    server ...
```

### Output of `haproxy -vv`

```plain
HAProxy version 2.6.5-987a4e2 2022/09/03 - https://haproxy.org/
Status: long-term supported branch - will stop receiving fixes around Q2 2027.
Known bugs: http://www.haproxy.org/bugs/bugs-2.6.5.html
Running on: Linux 5.4.0-122-generic #138~18.04.1-Ubuntu SMP Fri Jun 24 14:14:03 UTC 2022 x86_64
Build options :
  TARGET  = linux-glibc
  CPU     = generic
  CC      = cc
  CFLAGS  = -O2 -g -Wall -Wextra -Wundef -Wdeclaration-after-statement -Wfatal-errors -Wtype-limits -Wshift-negative-value -Wshift-overflow=2 -Wduplicated-cond -Wnull-dereference -fwrapv -Wno-address-of-packed-member -Wno-unused-label -Wno-sign-compare -Wno-unused-parameter -Wno-clobbered -Wno-missing-field-initializers -Wno-cast-function-type -Wno-string-plus-int -Wno-atomic-alignment
  OPTIONS = USE_PCRE2=1 USE_PCRE2_JIT=1 USE_STATIC_PCRE2=1 USE_OPENSSL=1 USE_LUA=1 USE_ZLIB=1 USE_DL=1
  DEBUG   = -DDEBUG_STRICT -DDEBUG_MEMORY_POOLS

Feature list : +EPOLL -KQUEUE +NETFILTER -PCRE -PCRE_JIT +PCRE2 +PCRE2_JIT +POLL +THREAD +BACKTRACE -STATIC_PCRE +STATIC_PCRE2 +TPROXY +LINUX_TPROXY +LINUX_SPLICE +LIBCRYPT +CRYPT_H -ENGINE +GETADDRINFO +OPENSSL +LUA +ACCEPT4 -CLOSEFROM +ZLIB -SLZ +CPU_AFFINITY +TFO +NS +DL +RT -DEVICEATLAS -51DEGREES -WURFL -SYSTEMD -OBSOLETE_LINKER +PRCTL -PROCCTL +THREAD_DUMP -EVPORTS -OT -QUIC -PROMEX -MEMORY_PROFILING

Default settings :
  bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with multi-threading support (MAX_THREADS=64, default=2).
Built with OpenSSL version : OpenSSL 1.1.1  11 Sep 2018
Running on OpenSSL version : OpenSSL 1.1.1  11 Sep 2018
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : TLSv1.0 TLSv1.1 TLSv1.2 TLSv1.3
Built with Lua version : Lua 5.4.4
Built with network namespace support.
Support for malloc_trim() is enabled.
Built with zlib version : 1.2.11
Running on zlib version : 1.2.11
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Built with PCRE2 version : 10.40 2022-04-14
PCRE2 library supports JIT : yes
Encrypted password support via crypt(3): yes
Built with gcc compiler version 7.5.0

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available multiplexer protocols :
(protocols marked as <default> cannot be specified using 'proto' keyword)
         h2 : mode=HTTP  side=FE|BE  mux=H2    flags=HTX|HOL_RISK|NO_UPG
       fcgi : mode=HTTP  side=BE     mux=FCGI  flags=HTX|HOL_RISK|NO_UPG
  <default> : mode=HTTP  side=FE|BE  mux=H1    flags=HTX
         h1 : mode=HTTP  side=FE|BE  mux=H1    flags=HTX|NO_UPG
  <default> : mode=TCP   side=FE|BE  mux=PASS  flags=
       none : mode=TCP   side=FE|BE  mux=PASS  flags=NO_UPG

Available services : none

Available filters :
        [CACHE] cache
        [COMP] compression
        [FCGI] fcgi-app
        [SPOE] spoe
        [TRACE] trace
```

### Last Outputs and Backtraces

```plain
[WARNING]  (...) : Server bbb is DOWN, reason: Socket error, check duration: 2ms. 2 active and 0 backup servers left. 0 sessions active, 0 requeued, 0 remaining in queue.
...
[WARNING]  (...) : Backup Server bbb is DOWN, reason: Socket error, check duration: 2ms. 1 active and 1 backup servers left. 0 sessions active, 0 requeued, 0 remaining in queue.
```

**Comment by wtarreau** [comment]
Thanks for your report, will have a look at it ASAP, sorry for the delay.

**Comment by wtarreau** [comment]
So a few points on this. First, if you get EPOLLHUP, it means the server closed, possibly with a response but not necessarily. Since it happened in response to a connect() and the FD was being polled for writes, it indicates the connect() call didn't succeed so the request wasn't sent. As such the response you're getting there is not the response to the health check but a failure reported by the server (or the server just closed because there were too many connections for it).

Regardless, haproxy should not return a misleading error here. It should accurately report what happened (and possibly any pending response the server might have sent, which could contain precious information about what the server was not happy with).

Your deep analysis of the sequence is particularly useful because, as you saw, it shows we're abusing FD_POLL_ERR. But that one must not be cleared (though in your test that was useful to confirm the cause), because a socket cannot be only temporarily in error, once in error, it's dead.

For me the problem here is that POLL_HUP was mapped to FD_EV_SHUT_RW since commit 6b3089856 ("MEDIUM: fd: do not use the FD_POLL_* flags in the pollers anymore") in 2.1, and I can't find any valid reason why. Oh yes I've just found  the reason why. I have a length test log here from the same date as the commit showing that when RDHUP is reported, then HUP is reported only if the sending side was previously closed. Even if that's true in practice with data transfers, it's a bad idea to make this shortcut because as you've seen, connect() does poll for writes but doesn't try to transfer anything hence and doesn't have to suffer from this shortcut.

Individually, none of these operations are fundamentally wrong, as they help determine a maximum of information from what's reported by the poller. But I do think that we should not set POLL_ERR when the output is closed while it's being polled for writes. The first reason is that this is a sticky flag and that it was not really reported by the lower layer here so we can be wrong (as was the case here), the second reason is that if it's reported we can expect getsockopt(SO_ERROR) to return an error and here it's not the case. As such, I think that the best solution is to undo a part of commit 2aaeee34da, that is, to remove the assignment of FD_POLL_ERR in fd_update_events() below the comment saying "SHUTW reported while FD was active for writes is an error", because this is the one that guesses wrong here. Anyway its sole purpose was to avoid a few send() errors on aborted connections, but these are totally marginal. And if we decided that we needed to reintroduce it, then we'd adopt distinct poll-for-writes mechanisms, with one indicating we're polling for a connect() instead.

As such I would appreciate it if you could test with this:
```
--- a/src/fd.c
+++ b/src/fd.c
@@ -587,10 +587,6 @@ int fd_update_events(int fd, uint evts)
              ((evts & FD_EV_SHUT_R)  ? FD_POLL_HUP : 0) |
              ((evts & FD_EV_ERR_RW)  ? FD_POLL_ERR : 0);
 
-       /* SHUTW reported while FD was active for writes is an error */
-       if ((fdtab[fd].state & FD_EV_ACTIVE_W) && (evts & FD_EV_SHUT_W))
-               new_flags |= FD_POLL_ERR;
-
        /* compute the inactive events reported late that must be stopped */
        must_stop = 0;
        if (unlikely(!fd_active(fd))) {
```

Thank you!

**Comment by gbrdead** [comment]
Thank you for the response and sorry for taking me so long to test your fix proposal.

I forgot to mention, but "Socket error" is not the only reason I see. There is one more that is very frequent:
**reason: Layer7 invalid response**

Your fix proposal does not fix the "Layer7 invalid response" reason. It fixes only the "Socket error" reason.
My pseudo-fix fixes both.

I sometimes get "Layer6 invalid response, info: SSL handshake failure (various errors here)" even with my pseudo-fix. However, these might be real - they are very infrequent and the way I reproduce the situation loads the CPU of the VM haproxy is running on. I am running haproxy itself and 10 concurrent curl's accessing a backend through haproxy at localhost.

**Comment by wtarreau** [comment]
Interesting. That's stranger then because it would indicate that either the error is set at another point, and/or that there are incorrect checks at some point. The problem with clearing the error is that it will be lost and may seriously result in stuck connections or such situations where an error event is lost. Errors should never be cleared. I'll have a check at sock_conn_check(), maybe it sets the error while there are still pending data and it should not do it.

**Comment by gbrdead** [comment]
I think I can explain this situation, too. Mind you, I haven't got the time to check the behaviour at runtime, I just analyzed the source. Here is a modification of the original description for the "Layer7 invalid response" case:

1. Unlike in the original description, let's assume that we receive only the following event:
EPOLLIN|EPOLLRDHUP
(and that we do not receive the one with EPOLLHUP).

2. As a result of EPOLLRDHUP, _do_poll() in ev_epoll.c calls fd_update_events() with FD_EV_SHUT_R set and FD_EV_SHUT_W unset. In the latter, FD_POLL_HUP gets set in fdtab[fd].state but FD_POLL_ERR stays unset.

3. In sock_conn_check() in sock.c we reach the comment:
/* error present, fall through common error check path */
(i.e., just like in the original description - we go to wait only if both FD_POLL_HUP and FD_POLL_ERR are unset).
This explains why my pseudo-fix works for this case, too.

...

5. raw_sock_to_buf() in raw_sock.c gets called.
fdtab[conn->handle.fd].state & (FD_POLL_ERR|FD_POLL_HUP)) == FD_POLL_HUP is true upon entry to the function (unlike in the original description) and we go to read0. The response from the backend is never read and haproxy assumes an application level error.

**Comment by wtarreau** [comment]
Thank you very much for your analysis. I don't like this, I'll have to recheck everything now. It's very hard for me to context-switch like this but this is needed. I think that the test on RDHUP that sets SHUT_R is a bit too early. Unless I'm mistaken, we only ought to report SHUT_R *after* we've read the last block (or HUP was reported without IN). Thus it definitely deserves a complete semantic recheck of what each flag promises and does not promise.

**Comment by gbrdead** [comment]
Are there any news about a fix?
Or is there a workaround, e.g. using a different I/O model (like the one in haproxy 1.8 which worked fine in our scenario)?

**Comment by wtarreau** [comment]
Sorry, but we're currently burried alive on the haproxyconf, it will be finished in a few days.

**Comment by wtarreau** [comment]
I'm getting progressively back on it again, now it's on the top of my pile of urgent stuff to anlayse before the release, thanks for your patience. It will take a while for me to re-digest the history (context-switching always being a nightmare). But I'd like to understand the whole sequence of events that causes this. Maybe it will require a much deeper fix that will not be suitable for a short-term but it needs to be sorted out.

**Comment by wtarreau** [comment]
Hi @gbrdead,

I'm getting a better understanding of what's happening. The FD_POLL_HUP internal status has a non-uniform meaning due to the fact that RDHUP exists on some pollers and not others. As such, it is set whenever a pending shut is reported by EPOLLRDHUP and normalized into FD_EV_SHUT_R. And indeed, in sock_conn_check() HUP is sufficient to declare a connect() error, so if the server returns data and immediately closes, this could trigger RDHUP, then SHUTR, then the connect error. I suspect that this patch would silence it (it's not the fix though):

diff --git a/src/fd.c b/src/fd.c
index f4f1bae81..a4ea96d77 100644
--- a/src/fd.c
+++ b/src/fd.c
@@ -584,7 +584,7 @@ int fd_update_events(int fd, uint evts)
        new_flags =
              ((evts & FD_EV_READY_R) ? FD_POLL_IN  : 0) |
              ((evts & FD_EV_READY_W) ? FD_POLL_OUT : 0) |
-             ((evts & FD_EV_SHUT_R)  ? FD_POLL_HUP : 0) |
+               //((evts & FD_EV_SHUT_R)  ? FD_POLL_HUP : 0) |
              ((evts & FD_EV_ERR_RW)  ? FD_POLL_ERR : 0);
 
        /* SHUTW reported while FD was active for writes is an error */

Another short-term alternative that should work would be to replace the `#ifndef EPOLLRDHUP` block in ev_epoll.c around line 32 with `#undef EPOLLRDHUP` then `#define EPOLLRDHUP 0`.

But clearly it indicates that the internal FD_POLL_HUP sometimes means POLLHUP and sometimes POLLRDHUP. I need to audit all of it now to understand when it implies a shutw and when not. I'll likely need to add an extra state but I would like to know if I should complement it or replace it.

**Comment by wtarreau** [comment]
Grrr sorry for the formating issue, let's try again:
```diff --git a/src/fd.c b/src/fd.c
index f4f1bae81..a4ea96d77 100644
--- a/src/fd.c
+++ b/src/fd.c
@@ -584,7 +584,7 @@ int fd_update_events(int fd, uint evts)
        new_flags =
              ((evts & FD_EV_READY_R) ? FD_POLL_IN  : 0) |
              ((evts & FD_EV_READY_W) ? FD_POLL_OUT : 0) |
-             ((evts & FD_EV_SHUT_R)  ? FD_POLL_HUP : 0) |
+               //((evts & FD_EV_SHUT_R)  ? FD_POLL_HUP : 0) |
              ((evts & FD_EV_ERR_RW)  ? FD_POLL_ERR : 0);
 
        /* SHUTW reported while FD was active for writes is an error */


```

**Comment by wtarreau** [comment]
There is also something else I don't get. The only way I'm seeing for epoll_wait() to report EPOLLHUP on a connect without EPOLLERR is that the port is really closed. Indeed, the only other case is when the server closes without reading the client's data, which will carry both EPOLLERR and EPOLLHUP. However here if we have something to send it will only happen after connect() succeeds, hence I can't see any way that connect() could be informed of destroyed data. And I've been spending quite some time trying to reproduce this, even with servers that send a response and immediately close without reading the request, and can't figure a way to get EPOLLHUP while waiting for connect().

It would help me a lot if you could share your strace output.

**Comment by wtarreau** [comment]
Not being able to reproduce it I can only guess, but my gut feeling is that in the end the patch below should be sufficient to address the problem and would be trivially backportable:
```
diff --git a/src/sock.c b/src/sock.c
index 1482c1a4f..2540d7d41 100644
--- a/src/sock.c
+++ b/src/sock.c
@@ -754,7 +754,7 @@ int sock_conn_check(struct connection *conn)
         * soon as we meet either of these delicate situations. Note that
         * SO_ERROR would clear the error after reporting it!
         */
-       if (cur_poller.flags & HAP_POLL_F_ERRHUP) {
+       if (0 && cur_poller.flags & HAP_POLL_F_ERRHUP) {
                /* modern poller, able to report ERR/HUP */
                if ((fdtab[fd].state & (FD_POLL_IN|FD_POLL_ERR|FD_POLL_HUP)) == FD_POLL_IN)
                        goto done;
```

It's the only place where we use HUP for what it really means (i.e. shut in both directions) and which corresponds to a connect() failure. However I'll really need your strace output, because the only way that code can cause a failure is by having the validating connect() return an unhandled errno, becausethe code above doesn't provoke an error, it just avoids double-checking connect() when it's known there was nothing suspicious. Thus if the code that follows reports the error, it's because errno doesn't match EALREADY, EINPROGRESS nor EISCONN.

**Comment by wtarreau** [comment]
I can, to some extents, provoke random POLLHUP + POLLERR just from the server, particularly if it doesn't consume the request or closes after disabling lineging. But these ones are reported on the recv() call via the h1 mux. Actually the whole response is valid but there's still the CO_FL_ERROR flag on the connection (since the error
was really present at the socket layer). There's still some ongoing work to make sure we don't use the connection errors from the upper layers anymore. Here for a test I could remove all of them and only rely on the stream's status, but that wasn't sufficient, some error cases were turned to timeouts, so probably a few places still rely on the hard error to interrupt some processing. But let's check for the strace output first before going into every direction.

**Comment by wtarreau** [comment]
For the specific case where the server responds without draining the client's data, causing random RSTs to be emitted, and random up/downs to happen, the following patch proposed by Christopher addresses it by making sure we offer a chance to the mux to first parse a full response before seeing the RST (it will only see the RST if the response is not complete). For me it totally stabilizes such checks in both 2.6 and 2.7, under load as well as when there is no load.

```diff
diff --git a/src/raw_sock.c b/src/raw_sock.c
index e172b1d4e..af95c8283 100644
--- a/src/raw_sock.c
+++ b/src/raw_sock.c
@@ -329,7 +329,7 @@ static size_t raw_sock_to_buf(struct connection *conn, void *xprt_ctx, struct bu
         * of recv()'s return value 0, so we have no way to tell there was
         * an error without checking.
         */
-       if (unlikely(fdtab[conn->handle.fd].state & FD_POLL_ERR))
+       if (unlikely(!done && fdtab[conn->handle.fd].state & FD_POLL_ERR))
                conn->flags |= CO_FL_ERROR | CO_FL_SOCK_RD_SH | CO_FL_SOCK_WR_SH;
        goto leave;
 }
```

**Comment by capflam** [comment]
FYI, I pushed the fix in 2.7-dev. I will be shipped with the 2.7-dev9.

**Comment by gbrdead** [comment]
Here are the requested strace logs. I used haproxy 2.6.6.

strace log for the situation I explained on Sep 13:
[socket_error.strace.log](https://github.com/haproxy/haproxy/files/10092434/socket_error.strace.log)

strace log for the situation I explained on Sep 27:
[layer7_invalid_response.strace.log](https://github.com/haproxy/haproxy/files/10092444/layer7_invalid_response.strace.log)

**Comment by gbrdead** [comment]
Now about the fix in 2.7-dev10:
It fixes only the situation I explained on Sep 27 - opposite to the fix proposed by [wtarreau](https://github.com/wtarreau) on Sep 17. Maybe you need to combine both fixes?

**Comment by wtarreau** [comment]
Many  thanks Vladimir! Your trace is amazing, it shows epoll_wait() reporting a hangup while this is not true, the subsequent send() works and the following recv() works as well to receive a response. That explains why your patch that masks these helps!

I have no idea how this is possible, for me it looks like epoll_wait() is lying. For example have you seen that recvfrom() after HUP+RDHUP which returns EAGAIN ? How is that possible at all ? It should return 0 indicating a recv shutdown. So it means we cannot trust these flags at all on your machine!

What exact kernel version are you using ? Because here the only possibility I'm seeing is to implement a workaround to ignore certain flags with a broken epoll implementation. I'd like to be able to detect this automatically, but we don't know how to trigger this situation though. But in the worst case we could imagine a global setting to enable such a workaround.

**Comment by gbrdead** [comment]
The kernel version is:
Linux 6749a9bb-63c7-4a1e-a968-1c1bd0d9266e 5.4.0-132-generic #148~18.04.1-Ubuntu SMP Mon Oct 24 20:41:14 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux

The VM is running under Cloud Foundry BOSH. The cloud infrastructure is AWS but I have been able to reproduce it on Azure, too (I can also check GCP, if you wish).

**Comment by wtarreau** [comment]
Thank you! There's no need to run in a different environment, the syscall is generic and shouldn't cause this. I cannot think of a single explanation for this polling state, but if it happens with such a generic kernel, we may need to implement a workaround.

Would it be possible to run tcpdump between haproxy and one server to see the exact sequence of network exchanges that produces this ? Do not hesitate to tell me if it's problematic to share such info. I'm trying to narrow down the issue to its exact root cause (maybe we can even reach a point where we could fix it in the kernel).

**Comment by wtarreau** [comment]
Just to make sure not to waste your time, the output of `tcpdump -Svvni ethX host $server_ip and tcp port $server_port` will be sufficient. Preferably on the working case (with your patch of Sep13) since it shows the subsequent syscalls working while they should not. Ideally with the output of `strace -tt` in parallel but that is not mandatory since we already have your trace :-)

**Comment by gbrdead** [comment]
It doesn't happen on my command. And if I apply my patch, I can no longer detect when it happens. Is it OK if I do it without the patch? The case in which the response is read (Socket error) is OK, isn't it?

**Comment by wtarreau** [comment]
Yeah makes sense of course. I thought you could reproduce it easily out of production. Do not hesitate to contact me privately for the trace in order not to share private info if you want.

**Comment by gbrdead** [comment]
Here is the "Socket error" case, both strace and tcpdump are in the same file.

[socket_error.log](https://github.com/haproxy/haproxy/files/10094416/socket_error.log)

**Comment by wtarreau** [comment]
Thank you, and good idea to have merged them into a single file: reordering them by time shows that epoll_wait() already responses with EPOLLIN|EPOLLHUP|EPOLLRDHUP as soon as the SYN-ACK comes back. The sequence of syscalls is so trivial that it should be reproducible. I do have access to a machine with the same distro and a reasonably similar kernel. I'll run some tests to try to reproduce it and figure the root cause. I may have to report it to netdev (thanks for having anonymized your traces, that will save me some time).

Now the next step might be to figure how to detect it at boot and fall back to a compatible mode. Oh by the way, could you please run with "haproxy -de" (or use `noepoll` in your global section) ? This will disable epoll and fall back to poll(). Both are expected to report the same set of flags. If it never fails with poll() we'll at least figure that the problem is with epoll. Otherwise it might be in the network stack.

**Comment by gbrdead** [comment]
I ran haproxy with poll (i.e. with -de) for a whole night and there was not a single reproduction. In the same situation, epoll gets the problem reproduced within minutes. So we can safely assume that poll is not affected.
In fact, I may even use poll. What is the downside, especially compared to haproxy 1.8?

**Comment by wtarreau** [comment]
OK great, we're getting closer!  The problem with poll() vs epoll() is that poll() will start to consume heavy CPU when you reach thousands of concurrent connections. But on lower loads it's not a problem at all.

**Comment by wtarreau** [comment]
Hello @gbrdead 

Could you please confirm that the server is a physically different machine here and not just a docker container or so ? I'm trying to set up a reproducer on a similar distro+kernel (for now I'm failing). Also could you please post the output of "lsmod" on your machine in case something gives me an idea ? For example we could imagine that one iptables rule or module is triggering the issue. Thanks!

**Comment by gbrdead** [comment]
The machine that haproxy is running on and the machine that the backend is running on are different virtual machines. I have no way of knowing whether they run on different physical machines but I guess this doesn't matter.

lsmod on the haproxy machine:
Module                  Size  Used by
binfmt_misc            24576  1
netlink_diag           16384  0
unix_diag              16384  0
xt_tcpudp              20480  4
iptable_mangle         16384  1
xt_cgroup              16384  3
bpfilter               24576  0
intel_rapl_msr         20480  0
intel_rapl_common      24576  1 intel_rapl_msr
crct10dif_pclmul       16384  1
crc32_pclmul           16384  0
ghash_clmulni_intel    16384  0
aesni_intel           372736  0
crypto_simd            16384  1 aesni_intel
cryptd                 24576  2 crypto_simd,ghash_clmulni_intel
glue_helper            16384  1 aesni_intel
cirrus                 16384  0
rapl                   20480  0
drm_kms_helper        184320  3 cirrus
drm                   495616  3 drm_kms_helper,cirrus
input_leds             16384  0
fb_sys_fops            16384  1 drm_kms_helper
syscopyarea            16384  1 drm_kms_helper
sysfillrect            16384  1 drm_kms_helper
sysimgblt              16384  1 drm_kms_helper
serio_raw              20480  0
mac_hid                16384  0
sch_fq_codel           20480  3
ip_tables              32768  1 iptable_mangle
x_tables               45056  4 xt_cgroup,xt_tcpudp,ip_tables,iptable_mangle
autofs4                45056  2
psmouse               151552  0
floppy                 81920  0
i2c_piix4              28672  0
ixgbevf                77824  0
pata_acpi              16384  0

**Comment by wtarreau** [comment]
OK thank you, that's useful. Not being in the same VM at least indicates that you're not dealing with local connections. Also nothing fancy in your modules at first glance.

**Comment by wtarreau** [comment]
Hi Vladimir,

there's still something that puzzles me in your trace. The second `connect()` that's used to verify that the connection was completely established should always return -1, either EISCONN to indicate that the connection is ready, EINPROGRESS to indicate it's still not ready, or any other error. Here it's returning 0, indicating an immediate, successful connect. It's as if the connection was silently aborted in the kernel and decided to instantly succeed the second time. This is what triggers the rest of the problem, as this is not supposed to happen. We're doing this instead of getsockopt(SO_ERROR) so that we don't flush the error queue that's needed by health checks. But I'm seeing how we could perform a getsockopt() and only skip it for checks. We might then get a little bit more info there about what's happening in such rare cases. But I still can't figure how it's possible to get two successful connect() calls on one single connection, nor how it's possible that the kernel side returns these flags upon a first connect().

I'll try to arrange that part and propose you a patch to test.

**Comment by wtarreau** [comment]
I just noticed it was marked as fixed while some extra work is still needed.

**Comment by wtarreau** [comment]
Hi @gbrdead,

I could finally prepare and test a patch that mkes sure to call connect() only once, and to use getsockopt() for the validation. This means that if connect() fails, we'll always get the error reported accurately in the checks. I'd like you to test it in your environment and see what it reports. If you're still seeing strange things, running strace again would be of great help. This time it will not be possible that a second connect() hides a spurious error, but since we already don't know how this sequence is possible at all, we should get prepared to seeing any surprise, including a getsockopt() returning no error or such funny things. At least it will make the checks and prod traffics more consistent (and we still don't know why regular traffic doesn't seem to experience any problem).

Thanks!
[0001-BUG-MEDIUM-connection-use-getsockopt-SO_ERROR-only-f.patch.txt](https://github.com/haproxy/haproxy/files/10332059/0001-BUG-MEDIUM-connection-use-getsockopt-SO_ERROR-only-f.patch.txt)

**Comment by gbrdead** [comment]
Hi, Willy.

1. haproxy 2.6.6: The error is easy to reproduce, both versions ("Socket error" and "Layer7 invalid response") are equally frequent. The last patch does not change anything.

2. haproxy 2.6.7: This release contains the linked change by Christopher Faulet. Only the "Layer7 invalid response" version reproduces but as frequently as in 2.6.6. The last patch does not change anything.

3. haproxy 2.8-dev0: This version behaves very differently from everything I've seen till now. The errors are much less frequent but they are also different:

- Socket error, info: "Broken pipe"
- Socket error (even though the change by Christopher is in)
"Layer7 invalid response" hasn't reproduced a single time yet.

The last patch does not change anything.
This version seems like an improvement over 2.6 (I don't know why) but the errors are not fully fixed. It also behaves differently - I am worried that a fix for 2.6 might not be a fix for 2.8 and vice versa.


In short, the patch does not change anything. Most probably, connect() and getsockopt(SO_ERROR) return the same thing.

**Comment by gbrdead** [comment]
About this:

> and we still don't know why regular traffic doesn't seem to experience any problem

My impression after the initial debugging was that the difference is in the way haproxy checks for errors its own data structures (I don't remember why exactly). The kernel, network hardware or whatever causes the unexpected HUPs most likely behaves in the same way. Also, remember that haproxy 1.8 did not suffer at all in the same environment.

**Comment by wtarreau** [comment]
Hi,

thanks for your tests. I would really need to see the strace showing what getsockopt() returns here. The point is that there's a major difference between the two. There's a bug somewhere down the stack making the first connect() fail and the second succeed despite the socket being in error. The network traces don't show anything wrong, which leaves us with no idea about what is causing this error. However, the trouble is that when we use the second connect, it finally manages to ressuscitate the socket which was already marked as dead, and that's absolutely not expected (nor does it correspond to a valid state transition). It's just hiding the dust under the carpet. And some of the previous attempts above only resulted in trying various hacks to try to hide the problem and be able to use that partially dead socket. But that's not correct. We *need* to report the real error (the one that every other machine reports on the second connect() and that getsockopt() will return without changing the connection's state). We may figure that the cause of this error will need some retries on the checks or whatever, I don't know. But we definitely need to report this failure because there is one reported by the OS and trying to hide it doesn't solve anything. Now as to why it differs between versions, sadly that's simple, it depends on the amount of error control placed at different key points. Here we're facing an impossible case, traffic flowing on a broken socket, so sometimes swapping two error checks to factor some error recovery code can simply allow something not expected to progress a bit further or to be caught a bit earlier.

**Comment by gbrdead** [comment]
Hi, Willy.

I used haproxy 2.6.7 with the patch applied.
layer7_invalid_response.strace.1.log is the strace from a reproduction. No call to getsockopt(SO_ERROR) is observed because _objt_check(((struct session *)conn->owner)->origin))_ is true.
So I commented out the block with the aforementioned check and got layer7_invalid_response.strace.2.log.

The situation is the same as in my explanation from 2022/09/22 - EPOLLRDHUP is received before the socket is connected, it gets ignored for a while, but eventually leads to a wrong decision that the backend is down in raw_sock_to_buf(), even though getsockopt(SO_ERROR) has returned no error in the mean time.

[layer7_invalid_response.strace.2.log](https://github.com/haproxy/haproxy/files/10428938/layer7_invalid_response.strace.2.log)
[layer7_invalid_response.strace.1.log](https://github.com/haproxy/haproxy/files/10428939/layer7_invalid_response.strace.1.log)

**Comment by wtarreau** [comment]
Thanks, was busy with other issues and backports. I'll have a look soon. At first glance the first connect() was enough to establish the connection. This is confirmed by sendto() alone working and getsockopt() returning success. Previously there was a doubt that it could have been the second connect() that "repaired" the connection and rendered it usable. I'll need to ask some kernel developers if they have an idea about that one because the problem is that what the poller reports doesn't match reality, and if we cannot trust the poller we're f*cked :-(

One short-term option I'm thinking about in this situation is to provide a global option to disable relying on EPOLLRDHUP, but at least I would like to know if we can document it as being needed on certain kernel versions only, and if possible to automatically detect it.

**Comment by wtarreau** [comment]
And in fact even then it would not be sufficient since you also provided reports showing EPOLLHUP previously :-(

**Comment by wtarreau** [comment]
Note that we could as well decide that we ignore both and always rely on recv()==0 like in the old days. That's an ugly workaround for a yet not-identified problem.

**Comment by wtarreau** [comment]
Please give this patch a try, it disables both HUP and RDHUP. It continues to work for me here. If we manage to find a way to automatically detect your system is affected by this problem we could even imagine doing that automatically. Detecting HUP after a connect() would be a bit hackish I guess, but could be an option.

[0001-WIP-epoll-disable-both-EPOLLHUP-and-EPOLLRDHUP.patch.txt](https://github.com/haproxy/haproxy/files/10468097/0001-WIP-epoll-disable-both-EPOLLHUP-and-EPOLLRDHUP.patch.txt)

**Comment by gbrdead** [comment]
Hi, Willy.

The patch works for 2.6.7. I ran my test setup for 12 hours and there was not a single reproduction. Without the patch, the issue reproduces about once per 10 minutes.
I can't tell anything about 2.8-dev2 - I was not able to reproduce any error without the patch.

**Comment by wtarreau** [comment]
Thanks for the test! This is becoming increasingly strange. I suspect that the work done by Christopher to centralize and postpone error reporting, combined with his previous patch for 2.7 contributed to helping here, but based on what we're seeing at the poller's level it remains concerning and possibly fragile.

I suspect the most reliable workaround we could think about would be to be able to mask HUP/RDHUP during the connect() phase. For now we don't know that there is a connect() phase at the FD level but I have to think about a way to deal with this. In the mean time it should be safe for you to keep that patch on top of your 2.6 if that helps (or alternately you may be interested in following 2.8's development).

**Comment by gbrdead** [comment]
If you do commit a fix (this or a different one), please link it to this issue. For now, I am not planning to use a custom patch.

**Comment by capflam** [comment]
Several versions were released. Any news on this issue ?

**Comment by gbrdead** [comment]
We are currently using 2.6.13 with the patch from 2023-01-20 and everything is OK.

**Comment by capflam** [comment]
What is the status for this issue ? Is it still relevant ?

**Comment by gbrdead** [comment]
As far as I know - yes. But we are using the patch from 2023-01-20 and the problem does not reproduce.
If you are asking whether the issue can be closed with no further action from your side - well, if no one else is experiencing the problem, I guess that the answer is "yes".

**Comment by wtarreau** [comment]
Thanks. So I think we should streamline that patch with an option to disable the use of EPOLLRDHUP so that it can simply end up as an option (ideally we could detect it but I'm not sure how).

**Comment by wtarreau** [comment]
I forgot that there's this dirty special case for HUP as well, so it's not just a matter of disabling RDHUP support :-( For this one I'm not sure how to proceed since it will affect all platforms, some more thinking is deserved. Maybe we'll end up with a configurable mask but I find this ugly :-/

**Comment by capflam** [comment]
A gentle ping to make the issue more visible :)

**Comment by capflam** [comment]
FYI, A fix was finally pushed. It was backported as far as 2.8 for now. But it should be backported as far as 2.4.
